canonical/cloud-init
# fix(wsl): Properly assemble multipart data

## Proposed Commit Message

```
fix(wsl): Properly assemble multipart data

In the case of Pro, if either agent or user data is not a cloud-config, we need
to set `self.userdata_raw` as if the user had written an #include file so
cloud-init transforms that internally into a multipart data.
We're passing strings and lists directly, which confused the processor due the
lack of a mime type.

Being explicit about only loading yaml for cloud-config files also allow other
composition of cloud-init features to just work, like jinja templates.

I only figured out this bad behavior when testing with empty Landscape
data, but any non-cloud-config content type would trigger the same behavior.
As we expect most of the time to be dealing with cloud-config, thus
dicts, (as the UP4W agent data format is) I'm special-casing empty user-data
to empty dict, which is easier to reason about than the multipart alternative.

```

## Additional Context
I only figured out this bad behavior when testing with empty Landscape data,
thus the test added for the case of empty Landscape data.
Any non-cloud-config content type would trigger the same erroneous behavior.
jibel showed me an interesting traceback due a syntax error in the user data file,
and that motivated investigating other edge use cases. I found the only way to
properly pass heterogenous pieces of userdata into cloud-init is by modelling
them as an `#include` file.
Unfortunately I couldn't figure out a proper way to assert on the rendering of
jinja templates in the unit test file, as it happens outside of the datasource.
The same is valid for the assumption that cloud-init will apply the agent data
in the presence of syntax errors in the user-data, as that decision is also taken
outside of the datasource, as far as I can tell.
This implementation is more robust and allows for other cloud-init features to
just work.

## Test Steps
To test that behavior:
1. Have a WSL instance with cloud-init 24.2 or later pre-packaged.
2. Set some agent.yaml cloud-config data and a non-cloud-config (empty or a shell script or json) user-data file matching the instance name to be set (such as `Ubuntu-24.04.user-data) in `%USERPROFILE%/.ubuntupro/.cloud-init`
3. Run cloud-init as usual.
4. It should error out and `cloud-init status --long` should show:

```
root@WinDev2407Eval:~# cloud-init status --long
status: error
extended_status: error - done
boot_status_code: enabled-by-generator
last_update: Thu, 01 Jan 1970 03:49:05 +0000
detail: DataSourceWSL
errors:
        - 'list' object has no attribute 'encode'
recoverable_errors:
ERROR:
        - failed stage init
```

"cloud-init.log" will contain a trace similar to:

```
2024-07-23 19:21:28,855 - main.py[ERROR]: failed stage init
Traceback (most recent call last):
File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 817, in status_wrapper
    ret = functor(name, args)
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 489, in main_init
    init.update()
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 556, in update
    self._store_rawdata(self.datasource.get_userdata_raw(), "userdata")
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 596, in _store_rawdata
    util.write_file(self._get_ipath("%s_raw" % datasource), data, 0o600)
  File "/usr/lib/python3/dist-packages/cloudinit/util.py", line 2290, in write_file
  content = encode_text(content)
                  ^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/cloudinit/util.py", line 154, in encode_text
  return text if isinstance(text, bytes) else text.encode(encoding=encoding)
                                               ^^^^^^^^^^^
AttributeError: 'list' object has no attribute 'encode'
```

Closes UDENG-3718

## Merge type
